### PR TITLE
Determine candidate ordering for Fair Sharing in Hierarchical Cohorts

### PR DESCRIPTION
**Pull request by gabesaba**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
We define the order in which to consider workloads
in the scheduling cycle, that is compatible with Fair Sharing in Hierarchical Cohorts (#3759).

We run a tournament, from leaf to root, where each node nominates
a workload which it considers greedily most fair. This process is repeated
until one workload remains at the root. See comments for detailed description of algorithm.

#### Special notes for your reviewer:
I wanted to get feedback on the algorithm before
investing time in tests. I will add those after initial pass.

#### Does this PR introduce a user-facing change?
No release note for this change, as we will do a single release note for the entire feature

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/fair_sharing_iterator.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+// DRS after admission of its nominated workload.
+func (ec *entryComparer) computeDRS(rootCohort *cache.CohortSnapshot, cqToEntry map[*cache.ClusterQueueSnapshot]*entry) {
+	for _, cq := range rootCohort.SubtreeClusterQueues() {
+		entry, ok := cqToEntry[cq]
+		if !ok {
+			continue
+		}
+		// We add workload's usage to CQ, so that all
+		// subsequent DRS include the admission of workload.
+		cq.AddUsage(entry.Usage())
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/fair_sharing_iterator.go:170]: I think it needs to encapsulate TAS. So in my PR I change the definition to do so: https://github.com/kubernetes-sigs/kueue/pull/4418/files#diff-fb3d6b36d3cd727e1dd2353fb86c0aee0bfaae1c900e0f005ab913620021d551R324-R328. So, I would suggest using this one if possible. We should also ultimately add tests for TAS + HC.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -199,7 +199,12 @@ func (s *Scheduler) schedule(ctx context.Context) wait.SpeedSignal {
 	entries := s.nominate(ctx, headWorkloads, snapshot)
 
 	// 4. Create iterator which returns ordered entries.
-	var iterator entryIterator = makeIterator(entries, s.workloadOrdering, s.fairSharing.Enable)
+	var iterator entryIterator
```
Comment (mimowo) [inline pkg/scheduler/scheduler.go:202]: nit, I would fold this dispatching into the preexisting `makeIterator` to offload the main scheduling function. 
The function is complex already, so minimizing "technical" distractions is preferred imo.

File: pkg/scheduler/fair_sharing_iterator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/fair_sharing_iterator.go:170]: Yeah, ideally we have one. I think conceptually for scheduling it makes sense to use the "assignment usage" `usage()` currently - coming from the actual assignement. I'm not yet sure why / if we need to use the Usage at all.

[the remark this answers is not in the document]
Comment (gabesaba) [reply pkg/scheduler/fair_sharing_iterator.go]: Done.

[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/fair_sharing_iterator.go]: IIUC for an admitted workload they are equivalent, but for an workload pending admission Info.Usage() may not be accurate as it sets "" for Flavors (as in the comment): . Since scheduler operates on entries which are not admitted yet, I think it shouldn't use Info.Usage(), but e.usage(). Opened https://github.com/kubernetes-sigs/kueue/pull/4510 to also rename the usage.

**Comment by mimowo** [comment]
"Pod groups when Single CQ should allow to preempt the lower priority group" this looks like a known flake: https://github.com/kubernetes-sigs/kueue/issues/4434
but "Metrics when workload is admitted with admission checks should ensure the admission check metrics are available" I haven't seen failing on other branches. Maybe it is a rare one off, but we should be careful here. Let me retry.

**Comment by mimowo** [comment]
/test pull-kueue-test-e2e-main-1-30

And just to have more data points:
/test pull-kueue-test-e2e-main-1-29
/test pull-kueue-test-e2e-main-1-31
/test pull-kueue-test-e2e-main-1-32

**Review by mimowo — COMMENTED** [review]
The code looks good, I will yet read it in more details, but I'm missing tests.

What is the unit test coverage of the `pkg/scheduler/fair_sharing_iterator.go`? I assume some of the code isn't covered.

So, I would suggest unit tests in scheduler_test.go - could be a dedicated test function or extending the existing, and one integration test. The integration test could be a follow up.

**Comment by mimowo** [comment]
To have more data points:
/test pull-kueue-test-e2e-main-1-29
/test pull-kueue-test-e2e-main-1-30
/test pull-kueue-test-e2e-main-1-31
/test pull-kueue-test-e2e-main-1-32

**Comment by gabesaba** [comment]
> The code looks good, I will yet read it in more details, but I'm missing tests.
> 
> What is the unit test coverage of the `pkg/scheduler/fair_sharing_iterator.go`? I assume some of the code isn't covered.
> 
> So, I would suggest unit tests in scheduler_test.go - could be a dedicated test function or extending the existing, and one integration test. The integration test could be a follow up.

See special note description :)

After adding tests, we hit every line, except for line 79 (return nil in getCq). This is expected: we should never hit this line, as pop() is only called when hasNext() -> true, implying a member of cqToEntry.

**Comment by gabesaba** [comment]
/hold

adding more tests

File: pkg/scheduler/fair_sharing_iterator.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	f.entryComparer.computeDRS(root, f.cqToEntry)
+	entry := runTournament(root, f.entryComparer, f.cqToEntry)
+
+	delete(f.cqToEntry, entry.clusterQueueSnapshot)
+	return entry
+}
+
+// getCq returns a CQ with a workload pending scheduling.
+func (f *fairSharingIterator) getCq() *cache.ClusterQueueSnapshot {
+	for cq := range f.cqToEntry {
```
Comment (mimowo) [inline pkg/scheduler/fair_sharing_iterator.go:76]: This picks up the "first" element from hashmap, but I believe this may depend on the PRNG: https://pkg.go.dev/hash/maphash. 

This is probably not a problem for the algorithm e2e, because IIUC the order of processing the root cohorts would not matter, as they use disjoint resource pools. 

Mostly I want to confirm my understanding, but if this is the case I would suggest to either add a comment why this is safe e2e or make sure a deterministic element is chosen.

File: pkg/scheduler/fair_sharing_iterator.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+		if !ok {
+			continue
+		}
+		// We add workload's usage to CQ, so that all
+		// subsequent DRS include the admission of workload.
+		cq.AddUsage(entry.usage())
+
+		// calculate DRS, with workload, for CQ.
+		dominantResourceShare := cq.DominantResourceShare()
+		ec.drsValues[drsKey{parentCohort: cq.Parent().GetName(), workloadKey: workload.Key(entry.Obj)}] = dominantResourceShare
```
Comment (mimowo) [inline pkg/scheduler/fair_sharing_iterator.go:178]: Do we need to consider this as a special case? Maybe this is shorter as:

```golang
rNode = cq
for rNode.HasParent() {
	dominantResourceShare := cohort.DominantResourceShare()
	ec.drsValues[drsKey{parentCohort: cohort.Parent().GetName(), workloadKey: workload.Key(entry.Obj)}] = dominantResourceShare
	rNode = rNode.Parent()
}
```

Comment (gabesaba) [reply pkg/scheduler/fair_sharing_iterator.go:178]: What is rNode's type? Is there a concise way to do this in go type system? Last time I did something similar, I had to define a new method to get the parent, as go's type system wasn't happy with

```go
type hierarchicalResourceNode interface {
     Parent() hierarchicalResourceNode
}
```

as Parent() returns some Cohort type, not the interface type. So instead I had to go with this hack, defining a method parentHRN which returns the interface type:

https://github.com/kubernetes-sigs/kueue/blob/92e194db4f47e54bddd05149175ef9052ac96c2d/pkg/cache/resource_node.go#L74-L79

File: pkg/scheduler/fair_sharing_iterator.go (hunk 3 above)
Comment (gabesaba) [reply pkg/scheduler/fair_sharing_iterator.go:76]: I'll add a comment. While the scheduling order is non-deterministic, scheduling is (nearly*) deterministic within any given Cohort. I don't think we should rely on scheduling order being deterministic outside of the Cohort scope

*edge case where DRF, priority, and timestamp are the same. We don't have a UUID tiebreak as we have in preemption

File: pkg/scheduler/fair_sharing_iterator.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	if !cq.HasParent() {
+		entry := f.cqToEntry[cq]
+		delete(f.cqToEntry, cq)
+		return entry
+	}
+
+	// CQ is part of a Cohort. We run a tournament, to select the
+	// most fair workload at each level.
+	root := cq.Parent().Root()
+	f.entryComparer.computeDRS(root, f.cqToEntry)
```
Comment (mimowo) [inline pkg/scheduler/fair_sharing_iterator.go:67]: I think it will be useful very useful for debuggablity to log `drsValues` at V(3) or V(5). I would suggest also to clear the "old" values. to avoid noise.

File: pkg/scheduler/fair_sharing_iterator.go (hunk 3 above)
Comment (mimowo) [reply pkg/scheduler/fair_sharing_iterator.go:76]: sgtm

**Comment by mimowo** [comment]
LGTM, left only small-ish comments.

> /hold
> adding more tests

More tests would be welcome, but I'm not considering them blocking as the current coverage is nearly 100%. Might be also a follow up to me. 

cc @tenzen-y you are welcome to review, but I think the PR is nearing the end, and I would like to merge it when the comments are addressed to avoid delaying the HC + FS preemptions, which are a follow up.

File: pkg/scheduler/fair_sharing_iterator.go (hunk 4 above)
Comment (mimowo) [reply pkg/scheduler/fair_sharing_iterator.go:178]: Another idea here is to have a function for CQ PathToRoot which returns a slice of cohorts or seq.Iter to the root. then we can just range over it.. In any case I'm ok with a follow up.

Comment (mimowo) [reply pkg/scheduler/fair_sharing_iterator.go:178]: sgtm

File: pkg/scheduler/fair_sharing_iterator.go (hunk 5 above)
Comment (gabesaba) [reply pkg/scheduler/fair_sharing_iterator.go:67]: Done.

**Comment by mimowo** [comment]
/test pull-kueue-test-e2e-main-1-30
Due to https://github.com/kubernetes-sigs/kueue/issues/4480

File: pkg/scheduler/fair_sharing_iterator.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+			cohort = cohort.Parent()
+		}
+
+		cq.RemoveUsage(entry.usage())
+	}
+}
+
+func (ec *entryComparer) clearDrsValues() {
+	for key := range ec.drsValues {
+		delete(ec.drsValues, key)
```
Comment (mimowo) [inline pkg/scheduler/fair_sharing_iterator.go:212]: non-blocking comment: This may actually result in re-allocating the map as it shrinks. I think just allocating a new map and letting the old be GCed it better.

File: pkg/scheduler/fair_sharing_iterator.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+func (f *fairSharingIterator) hasNext() bool {
+	return len(f.cqToEntry) > 0
+}
+
+func (f *fairSharingIterator) pop() *entry {
+	cq := f.getCq()
+
+	// CQ has no Cohort. We simply return its workload.
+	if !cq.HasParent() {
+		f.log.V(3).Info("Returning workload from ClusterQueue without Cohort", "clusterQueue", cq.GetName())
```
Comment (mimowo) [inline pkg/scheduler/fair_sharing_iterator.go:67]: non blocking nit: It could be helpful to put the workload name actually as avaiable so that when debugging the line would pop up when grepping by workload name

**Comment by mimowo** [comment]
/lgtm
/approve
@gabesaba I left two non-blocking comments. Feel free to address or release the hold and move on. We can always address them as follow ups.
/hold

**Comment by mimowo** [comment]
/unhold
Actually no need to wait, just feel fee to follow up post merge

**Comment by gabesaba** [comment]
Thanks for the review!

File: pkg/scheduler/fair_sharing_iterator.go (hunk 7 above)
Comment (gabesaba) [reply pkg/scheduler/fair_sharing_iterator.go:67]: https://github.com/kubernetes-sigs/kueue/pull/4517

File: pkg/scheduler/fair_sharing_iterator.go (hunk 6 above)
Comment (gabesaba) [reply pkg/scheduler/fair_sharing_iterator.go:212]: https://github.com/kubernetes-sigs/kueue/pull/4518

**Comment by tenzen-y** [comment]
> LGTM, left only small-ish comments.
> 
> > /hold
> > adding more tests
> 
> More tests would be welcome, but I'm not considering them blocking as the current coverage is nearly 100%. Might be also a follow up to me.
> 
> cc @tenzen-y you are welcome to review, but I think the PR is nearing the end, and I would like to merge it when the comments are addressed to avoid delaying the HC + FS preemptions, which are a follow up.

I agree with you. I will review this later. And leave comments or address those by myself.

**Comment by gabesaba** [comment]
/release-note-edit
```release-note
[FSxHC] Make Fair Sharing compatible with Hierarchical Cohorts during scheduling
```
